v2/modules/jrl.cmake

```
@@ -0,0 +1,1957 @@
+    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin CACHE PATH "") # For Unix/MacOS executables, Windows: .exe, .dll, .pyd
+    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib CACHE PATH "") # for Unix/MacOS shared libraries .so/.dylib and Windows: .lib (import libraries for shared libraries)
+    set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib CACHE PATH "") # For static libraries add_library(STATIC ...) .a and Windows: .lib
+
+    # /!\ MODULE libraries are dynamic libraries. On Windows, python modules are MODULE libraries, with pyd extension.
+    #     They should be placed explicitely in lib/site-packages when building python extensions.
+    foreach(config Debug Release MinSizeRel RelWithDebInfo)
+        string(TOUPPER ${config} config)
+        set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_${config} ${CMAKE_BINARY_DIR}/bin CACHE PATH "")
+        set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_${config} ${CMAKE_BINARY_DIR}/lib CACHE PATH "")
+        set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_${config} ${CMAKE_BINARY_DIR}/lib CACHE PATH "")
+    endforeach()
+endfunction()
+
+# jrl_target_set_output_directory(<target_name> OUTPUT_DIRECTORY <dir>)
+# This function configures the `ARCHIVE_OUTPUT_DIRECTORY`,
+# `LIBRARY_OUTPUT_DIRECTORY`, and `RUNTIME_OUTPUT_DIRECTORY` properties
+# for the specified target.
+# This is useful for python modules that need to be placed in a specific directory.
+# In this module we use it to place python modules in ${CMAKE_BINARY_DIR}/lib/site-packages
+# To mimic the installation layout.
+# Otherwise, the python modules being a RUMTIME target, would be placed in ${CMAKE_BINARY_DIR}/bin
+# Example:
+# ```cmake
+#   jrl_target_set_output_directory(my_python_module_target OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib/site-packages)
+# ```
+function(jrl_target_set_output_directory target_name)
+    set(options)
+    set(oneValueArgs OUTPUT_DIRECTORY)
+    set(multiValueArgs)
+    cmake_parse_arguments(arg "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+    jrl_check_target_exists(${target_name})
+    jrl_check_var_defined(arg_OUTPUT_DIRECTORY)
+
+    set(dir ${arg_OUTPUT_DIRECTORY})
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+            RUNTIME_OUTPUT_DIRECTORY ${dir}
+            LIBRARY_OUTPUT_DIRECTORY ${dir}
+            ARCHIVE_OUTPUT_DIRECTORY ${dir}
+    )
+
+    foreach(config Debug Release MinSizeRel RelWithDebInfo)
+        string(TOUPPER ${config} config)
+        set_target_properties(
+            ${target_name}
+            PROPERTIES
+                RUNTIME_OUTPUT_DIRECTORY_${config} ${dir}
+                LIBRARY_OUTPUT_DIRECTORY_${config} ${dir}
+                ARCHIVE_OUTPUT_DIRECTORY_${config} ${dir}
+        )
+    endforeach()
+endfunction()
+
+# Configures the default install directories using GNUInstallDirs (bin, lib, include, etc.)
+# Works on all platforms
+function(jrl_configure_default_install_dirs)
+    include(GNUInstallDirs)
+endfunction()
```

```suggestion
macro(jrl_configure_default_install_dirs)
    include(GNUInstallDirs)
endmacro()
```

Some variables like `${CMAKE_INSTALL_FULL_BINDIR}` are not defined if the include is inside a function.